<a href="https://colab.research.google.com/github/Ebaghae/DSVIICODE/blob/main/Data_Ingestion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ****Read csv File Using Pandas****

In [ ]:
import pandas as pd
Yellow_Trip = pd.read_csv('/content/yellow_tripdata_2016-03.csv')

The runtime for reading this files using pandas was 1
second.

In [ ]:
Yellow_Trip.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2016-03-01 00:00:00   2016-03-01 00:07:55              1.0   
1         1  2016-03-01 00:00:00   2016-03-01 00:11:06              1.0   
2         2  2016-03-01 00:00:00   2016-03-01 00:31:06              2.0   
3         2  2016-03-01 00:00:00   2016-03-01 00:00:00              3.0   
4         2  2016-03-01 00:00:00   2016-03-01 00:00:00              5.0   

   trip_distance  pickup_longitude  pickup_latitude  RatecodeID  \
0           2.50        -73.976746        40.765152         1.0   
1           2.90        -73.983482        40.767925         1.0   
2          19.98        -73.782021        40.644810         1.0   
3          10.78        -73.863419        40.769814         1.0   
4          30.43        -73.971741        40.792183         3.0   

  store_and_fwd_flag  dropoff_longitude  dropoff_latitude  payment_type  \
0                  N         -74.004265         40.746128           1.0   
1                  N         -74.005943         40.733166           1.0   
2                  N         -73.974541         40.675770           1.0   
3                  N         -73.969650         40.757767           1.0   
4                  N         -74.177170         40.695053           1.0   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0          9.0    0.5      0.5        2.05          0.00   
1         11.0    0.5      0.5        3.05          0.00   
2         54.5    0.5      0.5        8.00          0.00   
3         31.5    0.0      0.5        3.78          5.54   
4         98.0    0.0      0.0        0.00         15.50   

   improvement_surcharge  total_amount  
0                    0.3         12.35  
1                    0.3         15.35  
2                    0.3         63.80  
3                    0.3         41.62  
4                    0.3        113.80

In [ ]:
Yellow_Trip.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180713 entries, 0 to 180712
Data columns (total 19 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   VendorID               180713 non-null  int64  
 1   tpep_pickup_datetime   180713 non-null  object 
 2   tpep_dropoff_datetime  180712 non-null  object 
 3   passenger_count        180712 non-null  float64
 4   trip_distance          180712 non-null  float64
 5   pickup_longitude       180712 non-null  float64
 6   pickup_latitude        180712 non-null  float64
 7   RatecodeID             180712 non-null  float64
 8   store_and_fwd_flag     180712 non-null  object 
 9   dropoff_longitude      180712 non-null  float64
 10  dropoff_latitude       180712 non-null  float64
 11  payment_type           180712 non-null  float64
 12  fare_amount            180712 non-null  float64
 13  extra                  180712 non-null  float64
 14  mta_tax                180712 non-nu

# **Reading csv File Using Dask**

In [ ]:
import dask.dataframe as dd

In [ ]:
Yellow_Trip1 = dd.read_csv("/content/yellow_tripdata_2016-03.csv")

The runtime to read the csv file using dask is 0 seconds. This shows that Dask is faster in reading large files compared to pandas.

# **Data Validation**

In [ ]:
Yellow_Trip.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2016-03-01 00:00:00   2016-03-01 00:07:55              1.0   
1         1  2016-03-01 00:00:00   2016-03-01 00:11:06              1.0   
2         2  2016-03-01 00:00:00   2016-03-01 00:31:06              2.0   
3         2  2016-03-01 00:00:00   2016-03-01 00:00:00              3.0   
4         2  2016-03-01 00:00:00   2016-03-01 00:00:00              5.0   

   trip_distance  pickup_longitude  pickup_latitude  RatecodeID  \
0           2.50        -73.976746        40.765152         1.0   
1           2.90        -73.983482        40.767925         1.0   
2          19.98        -73.782021        40.644810         1.0   
3          10.78        -73.863419        40.769814         1.0   
4          30.43        -73.971741        40.792183         3.0   

  store_and_fwd_flag  dropoff_longitude  dropoff_latitude  payment_type  \
0                  N         -74.004265         40.746128           1.0   
1                  N         -74.005943         40.733166           1.0   
2                  N         -73.974541         40.675770           1.0   
3                  N         -73.969650         40.757767           1.0   
4                  N         -74.177170         40.695053           1.0   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0          9.0    0.5      0.5        2.05          0.00   
1         11.0    0.5      0.5        3.05          0.00   
2         54.5    0.5      0.5        8.00          0.00   
3         31.5    0.0      0.5        3.78          5.54   
4         98.0    0.0      0.0        0.00         15.50   

   improvement_surcharge  total_amount  
0                    0.3         12.35  
1                    0.3         15.35  
2                    0.3         63.80  
3                    0.3         41.62  
4                    0.3        113.80

In [ ]:
# Remove special character
Yellow_Trip.columns = Yellow_Trip.columns.str.replace(' ', '')
  

In [ ]:
#Rename columns
Yellow_Trip.rename(columns={'RatecodeID': 'ratecodeID', 'VendorID': 'vendorID'})

vendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0              1  2016-03-01 00:00:00   2016-03-01 00:07:55              1.0   
1              1  2016-03-01 00:00:00   2016-03-01 00:11:06              1.0   
2              2  2016-03-01 00:00:00   2016-03-01 00:31:06              2.0   
3              2  2016-03-01 00:00:00   2016-03-01 00:00:00              3.0   
4              2  2016-03-01 00:00:00   2016-03-01 00:00:00              5.0   
...          ...                  ...                   ...              ...   
180708         2  2016-03-01 10:23:37   2016-03-01 10:37:36              1.0   
180709         2  2016-03-01 10:23:37   2016-03-01 10:34:06              6.0   
180710         2  2016-03-01 10:23:37   2016-03-01 10:32:13              1.0   
180711         1  2016-03-01 10:23:38   2016-03-01 10:43:24              1.0   
180712         2  2016-03-01 10:23:38                   NaN              NaN   

        trip_distance  pickup_longitude  pickup_latitude  ratecodeID  \
0                2.50        -73.976746        40.765152         1.0   
1                2.90        -73.983482        40.767925         1.0   
2               19.98        -73.782021        40.644810         1.0   
3               10.78        -73.863419        40.769814         1.0   
4               30.43        -73.971741        40.792183         3.0   
...               ...               ...              ...         ...   
180708           1.49        -73.965446        40.774719         1.0   
180709           1.87        -73.955872        40.782269         1.0   
180710           0.84        -73.984756        40.744240         1.0   
180711           3.30        -74.008316        40.704338         1.0   
180712            NaN               NaN              NaN         NaN   

       store_and_fwd_flag  dropoff_longitude  dropoff_latitude  payment_type  \
0                       N         -74.004265         40.746128           1.0   
1                       N         -74.005943         40.733166           1.0   
2                       N         -73.974541         40.675770           1.0   
3                       N         -73.969650         40.757767           1.0   
4                       N         -74.177170         40.695053           1.0   
...                   ...                ...               ...           ...   
180708                  N         -73.968842         40.758469           1.0   
180709                  N         -73.960289         40.801579           2.0   
180710                  N         -73.977943         40.754719           1.0   
180711                  N         -73.994896         40.740116           1.0   
180712                NaN                NaN               NaN           NaN   

        fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0               9.0    0.5      0.5        2.05          0.00   
1              11.0    0.5      0.5        3.05          0.00   
2              54.5    0.5      0.5        8.00          0.00   
3              31.5    0.0      0.5        3.78          5.54   
4              98.0    0.0      0.0        0.00         15.50   
...             ...    ...      ...         ...           ...   
180708         10.0    0.0      0.5        2.16          0.00   
180709          9.5    0.0      0.5        0.00          0.00   
180710          7.0    0.0      0.5        1.56          0.00   
180711         14.5    0.0      0.5        3.05          0.00   
180712          NaN    NaN      NaN         NaN           NaN   

        improvement_surcharge  total_amount  
0                         0.3         12.35  
1                         0.3         15.35  
2                         0.3         63.80  
3                         0.3         41.62  
4                         0.3        113.80  
...                       ...           ...  
180708                    0.3         12.96  
180709                    0.3         10.30  
180710                    

In [ ]:
#Convert 2 columns data type from object to datetime.
Yellow_Trip[["tpep_dropoff_datetime", "tpep_pickup_datetime"]] = Yellow_Trip[["tpep_dropoff_datetime", "tpep_pickup_datetime"]].apply(pd.to_datetime)

In [ ]:
print(Yellow_Trip[["tpep_dropoff_datetime", "tpep_pickup_datetime"]].info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180713 entries, 0 to 180712
Data columns (total 2 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   tpep_dropoff_datetime  180712 non-null  datetime64[ns]
 1   tpep_pickup_datetime   180713 non-null  datetime64[ns]
dtypes: datetime64[ns](2)
memory usage: 2.8 MB
None


##**Write Utility File** 

In [ ]:
%%writefile utilityfile.py
import logging
import os
import subprocess
import yaml
import pandas as pd
import datetime 
import gc
import re


################
# File Reading #
################

def read_config_file(filepath):
    with open(filepath, 'r') as stream:
        try:
            return yaml.safe_load(stream)
        except yaml.YAMLError as exc:
            logging.error(exc)


def replacer(string, char):
    pattern = char + '{2,}'
    string = re.sub(pattern, char, string) 
    return string

def col_header_val(df,table_config):
    '''
    replace whitespaces in the column
    and standardized column names
    '''
    df.columns = df.columns.str.lower()
    df.columns = df.columns.str.replace('[^\w]','_',regex=True)
    df.columns = list(map(lambda x: x.strip('_'), list(df.columns)))
    df.columns = list(map(lambda x: replacer(x,'_'), list(df.columns)))
    expected_col = list(map(lambda x: x.lower(),  table_config['columns']))
    expected_col.sort()
    df.columns =list(map(lambda x: x.lower(), list(df.columns)))
    df = df.reindex(sorted(df.columns), axis=1)
    if len(df.columns) == len(expected_col) and list(expected_col)  == list(df.columns):
        print("column name and column length validation passed")
        return 1
    else:
        print("column name and column length validation failed")
        mismatched_columns_file = list(set(df.columns).difference(expected_col))
        print("Following File columns are not in the YAML file",mismatched_columns_file)
        missing_YAML_file = list(set(expected_col).difference(df.columns))
        print("Following YAML columns are not in the file uploaded",missing_YAML_file)
        logging.info(f'df columns: {df.columns}')
        logging.info(f'expected columns: {expected_col}')
        return 0

Writing utilityfile.py


## **Write YAML File**

In [ ]:
%%writefile file.yaml
file_type: csv
file_name: yellow_tripdata_2016-03
inbound_delimiter: ","
outbound_delimiter: "|"
skip_leading_rows: 1
columns: 
    - vendorID
    - tpep_pickup_datetime
    - tpep_dropoff_datetime
    - passenger_count
    - trip_distance
    - pickup_longitude
    - pickup_latitude
    - ratecodeID
    - store_and_fwd_flag
    - dropoff_longitude
    - dropoff_latitude
    - payment_type
    - fare_amount
    - extra
    - mta_tax
    - tip_amount
    - tolls_amount
    - improvement_surcharge
    - total_amount


Writing file.yaml


In [ ]:
# Read Configuration file
import utilityfile as util
config_data = util.read_config_file("file.yaml")

In [ ]:
#Check yaml file
config_data

{'file_type': 'csv',
 'file_name': 'yellow_tripdata_2016-03',
 'inbound_delimiter': ',',
 'outbound_delimiter': '|',
 'skip_leading_rows': 1,
 'columns': ['vendorID',
  'tpep_pickup_datetime',
  'tpep_dropoff_datetime',
  'passenger_count',
  'trip_distance',
  'pickup_longitude',
  'pickup_latitude',
  'ratecodeID',
  'store_and_fwd_flag',
  'dropoff_longitude',
  'dropoff_latitude',
  'payment_type',
  'fare_amount',
  'extra',
  'mta_tax',
  'tip_amount',
  'tolls_amount',
  'improvement_surcharge',
  'total_amount']}

In [ ]:
# read the file using config file
file_type = config_data['file_type']
source_file = "./" + config_data['file_name'] + f'.{file_type}'
#print("",source_file)
Data = pd.read_csv(source_file,config_data['inbound_delimiter'])
Data.head()

<ipython-input-15-2f175c70f79a>:5: FutureWarning: In a future version of pandas all arguments of read_csv except for the argument 'filepath_or_buffer' will be keyword-only.
  Data = pd.read_csv(source_file,config_data['inbound_delimiter'])


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2016-03-01 00:00:00   2016-03-01 00:07:55                1   
1         1  2016-03-01 00:00:00   2016-03-01 00:11:06                1   
2         2  2016-03-01 00:00:00   2016-03-01 00:31:06                2   
3         2  2016-03-01 00:00:00   2016-03-01 00:00:00                3   
4         2  2016-03-01 00:00:00   2016-03-01 00:00:00                5   

   trip_distance  pickup_longitude  pickup_latitude  RatecodeID  \
0           2.50        -73.976746        40.765152         1.0   
1           2.90        -73.983482        40.767925         1.0   
2          19.98        -73.782021        40.644810         1.0   
3          10.78        -73.863419        40.769814         1.0   
4          30.43        -73.971741        40.792183         3.0   

  store_and_fwd_flag  dropoff_longitude  dropoff_latitude  payment_type  \
0                  N         -74.004265         40.746128           1.0   
1                  N         -74.005943         40.733166           1.0   
2                  N         -73.974541         40.675770           1.0   
3                  N         -73.969650         40.757767           1.0   
4                  N         -74.177170         40.695053           1.0   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0          9.0    0.5      0.5        2.05          0.00   
1         11.0    0.5      0.5        3.05          0.00   
2         54.5    0.5      0.5        8.00          0.00   
3         31.5    0.0      0.5        3.78          5.54   
4         98.0    0.0      0.0        0.00         15.50   

   improvement_surcharge  total_amount  
0                    0.3         12.35  
1                    0.3         15.35  
2                    0.3         63.80  
3                    0.3         41.62  
4                    0.3        113.80

###**Validate Column Number and Names with Yaml File**

In [ ]:
util.col_header_val(Data,config_data)

column name and column length validation passed


1

### **Write the file in pipe separated text file (|) in gz format.**

In [ ]:
#Write file to text format with seperator - |
Yellow_Trip.to_csv('Yellow_Trip.txt', sep='|', index=False)

In [19]:
# Compress file into gzip format
from pandas.io.common import gzip
import shutil
with open('/content/Yellow_Trip.txt', 'rb') as f_in:
    with gzip.open('/content/Yellow_Trip.txt.gz', 'wb') as f_out:
        shutil.copyfileobj(f_in, f_out)


In [20]:
# Read compressed file
with gzip.open('/content/Yellow_Trip.txt.gz', 'rb') as f:
    file_content = f.read()

## **File Summary**


*   Total number of rows: 180713
*   Total number of columns: 19

*   File name: Yellow_Trip.txt.gz
*   File size: 2GB






